File: focus_todo_computer/focus-todo/src/sync/mockSyncFlow.ts
import type { Task, TaskSyncMetadata } from '../types/task'
import {
  syncCoordinator,
  type SyncCoordinatorState,
} from './syncCoordinator'
import type { CanonicalTask } from './canonicalTaskMapper'

export interface MockSyncCycleResult {
  outboundBatch: CanonicalTask[]
  coordinatorState: SyncCoordinatorState
}

export function initializeMockSyncState(input: {
  tasks: Task[]
  syncMetadata: TaskSyncMetadata[]
}): SyncCoordinatorState {
  return syncCoordinator.initializeAfterHydration(input)
}

export function prepareMockOutboundPreview(
  coordinatorState: SyncCoordinatorState,
): CanonicalTask[] {
  return syncCoordinator.prepareOutboundBatch(coordinatorState)
}

export function applyMockInboundTasks(
  coordinatorState: SyncCoordinatorState,
  mockData: CanonicalTask[],
): SyncCoordinatorState {
  return syncCoordinator.applyInboundBatch(coordinatorState, mockData)
}

export function runMockSyncCycle(
  coordinatorState: SyncCoordinatorState,
  mockData: CanonicalTask[] = [],
): MockSyncCycleResult {
  const outboundBatch = prepareMockOutboundPreview(coordinatorState)
  const nextCoordinatorState = applyMockInboundTasks(coordinatorState, mockData)

  return {
    outboundBatch,
    coordinatorState: nextCoordinatorState,
  }
}

File: focus_todo_mobile/lib/sync/mock_sync_flow.dart
import '../shared/models/task.dart';
import '../shared/models/task_sync_metadata.dart';
import 'canonical_task_mapper.dart';
import 'sync_coordinator.dart';

class MockSyncCycleResult {
  const MockSyncCycleResult({
    required this.outboundBatch,
    required this.coordinatorState,
  });

  final List<CanonicalTask> outboundBatch;
  final SyncCoordinatorState coordinatorState;
}

const MockSyncCoordinator _mockSyncCoordinator = MockSyncCoordinator();

SyncCoordinatorState initializeMockSyncState({
  required List<Task> tasks,
  required List<TaskSyncMetadata> syncMetadata,
}) {
  return _mockSyncCoordinator.initializeAfterHydration(
    tasks: tasks,
    syncMetadata: syncMetadata,
  );
}

List<CanonicalTask> prepareMockOutboundPreview(
  SyncCoordinatorState coordinatorState,
) {
  return _mockSyncCoordinator.prepareOutboundBatch(coordinatorState);
}

SyncCoordinatorState applyMockInboundTasks(
  SyncCoordinatorState coordinatorState,
  List<CanonicalTask> mockData,
) {
  return _mockSyncCoordinator.applyInboundBatch(coordinatorState, mockData);
}

MockSyncCycleResult runMockSyncCycle(
  SyncCoordinatorState coordinatorState, {
  List<CanonicalTask> mockData = const <CanonicalTask>[],
}) {
  final outboundBatch = prepareMockOutboundPreview(coordinatorState);
  final nextCoordinatorState = applyMockInboundTasks(
    coordinatorState,
    mockData,
  );

  return MockSyncCycleResult(
    outboundBatch: outboundBatch,
    coordinatorState: nextCoordinatorState,
  );
}
